lib/awful/spawn.lua

```diff
@@ -435,6 +470,149 @@ function spawn.read_lines(input_stream, line_callback, done_callback, close)
     start_read()
 end
 
+-- When a command should only be executed once or only have a single instance,
+-- track the SNID set on them to prevent multiple execution.
+spawn.single_instance_manager = {
+    by_snid = {},
+    by_uid  = {},
+}
+
+aclient.property.persist("single_instance_id", "string")
+
+-- Check if the client is running either using the rule source or the matcher.
+local function is_running(hash, matcher)
+    local status = spawn.single_instance_manager.by_uid[hash]
+    if not status then return false end
+
+    if #status.instances == 0 then return false end
+
+    for _, c in ipairs(status.instances) do
+        if c.valid then return true end
+    end
+
+    if matcher then
+        for _, c in ipairs(client.get()) do
+            if matcher(c) then return true end
+        end
+    end
+
+    return false
+end
+
+-- Keep the data related to this hash.
+local function register_common(hash, rules, callback)
+    if spawn.single_instance_manager.by_uid[hash] then return end
+
+    local status = {
+        rules     = rules,
+        callback  = callback,
+        instances = setmetatable({}, {__mode = "v"}),
+        hash      = hash,
+        exec      = false,
+    }
+
+    spawn.single_instance_manager.by_uid[hash] = status
+
+    return status
+end
+
+local function run_once_common(hash, cmd)
+    local ret, snid = spawn.spawn(cmd)
+
+    if not snid then return false end
+
+    assert(spawn.single_instance_manager.by_uid[hash])
+
+    local status = spawn.single_instance_manager.by_uid[hash]
+    status.exec = true
+
+    spawn.single_instance_manager.by_snid[snid] = status
+
+    return ret, snid
+end
+
+--- Spawn a command if it has not been spawned before.
+--
+-- This function tries its best to preserve the state across `awesome.restart()`.
+--
+-- By default, when no `uid` is specified, this function will generate one by
+-- hashing the command and its rules. If you have multiple instance of the same
+-- command and rules, you need to specify an UID or only the first one will be
+-- executed.
+--
+-- The `rules` are standard `awful.rules`.
+--
+-- This function depends on the startup notification protocol to be correctly
+-- implemented by the command. See `client.startup_id` for more information.
+-- Note that this also wont work with shell or terminal commands.
+--
+-- @tparam string|table cmd The command.
+-- @tparam table rules The properties that need to be applied to the client.
+-- @tparam[opt] string uid A string to identify the client so it isn't executed
+--  multiple time.
```

Comment:
I'd make it more explicit that `uid` is `unique id`. At first I was thinking of `user id`.